WT-7514 Let tiered subsystem open files on behalf of block manager.

ddanderson opened the pull request:
I've based this branch off of wt-7499 changes, which have not yet been merged.
It shows how I expect the block manager to open files based on their object id.
This does not yet addressed flush_tier changes for the block manager.

ddanderson commented:
@sueloverso, @keitharnoldsmith I've made this a draft since it's based off the wt-7499 (unmerged) pull request.  Here's the direction I'm going with the block manager to let the tier code have complete knowledge of how to name files and open them.

sueloverso commented on src/tiered/tiered_opener.c:
```diff
@@ -0,0 +1,75 @@
+
+    WT_UNUSED(opener);
+    tiered = (WT_TIERED *)cookie;
+
```
I don't know if you looked at it but the function that exists that you might want to use (maybe with some small changes) is `tiered/tiered_handle.c:__wt_tiered_name`.

ddanderson replied:
Done.

sueloverso commented on src/tiered/tiered_handle.c:
```diff
@@ -361,10 +377,37 @@ __tiered_switch(WT_SESSION_IMPL *session, const char *config)
+        prev_filename = prev_uri;
+        WT_PREFIX_SKIP_REQUIRED(session, prev_filename, "file:");
+        new_object_name = new_object_uri;
+        WT_PREFIX_SKIP_REQUIRED(session, new_object_name, "object:");
```
I think we may want a new no-prefix name option to `wt_tiered_name`, maybe `WT_TIERED_NAME_ONLY` that doesn't prepend the prefix. With the changes I merged today for the new suffix, `prev_filename` and `new_object_name` should be identical always. The same object names and the prefix indicates where it lives. Does that mean `ss_flush*` should only have one name arg instead of two?

sueloverso commented on src/tiered/tiered_handle.c:
```diff
@@ -361,10 +377,37 @@ __tiered_switch(WT_SESSION_IMPL *session, const char *config)
+
+    if (tiered->current_id > 1 && tiered->current_id != prev_id) {
+        /*
+         * This code should be done in its own thread, as it may incur network delays or failure.
```
I pushed a draft branch of the thread and work unit stuff. It is still very much in progress but I'll move this code into it when I write the function to process a flush work unit.

sueloverso commented:
I took a very quick look at this and left a couple comments.

ddanderson commented:
@sueloverso and @keitharnoldsmith, I made some changes to treat `WT_DHANDLE_TYPE_TIER` as a btree in several more cases, that has resolved the "clean close connection" issue.  The tiered tests run cleanly.  If I add in a connection reopen to the test, it does not work yet, that's next.

keitharnoldsmith reviewed (commented):
Nice work @ddanderson!

Glad you found that `WT_DHANDLE_TYPE_BTREE` check in `__wt_checkpoint_get_handles()`. That was probably important for getting data properly written.

ddanderson commented:
@sueloverso and @keitharnoldsmith, this branch is ready to review again.  The principal goals of this PR are:

- advance the tests so that we can do `flush_tier` and see objects move correctly (e.g. into the `mybucket` and `cache-mybucket` directory, for test_tiered02.py).
- Identify some limits of testing (new tickets WT-7589, WT-7591) so it's clear what to work on next to extend the work.
- Identify some other code deficiencies with other new tickets (WT-7588, WT-7590)
- remove unused/dead code (some vestiges from LABS-19, like tracking the current object id in the block manager, etc.)

Some of the more significant parts of this change:

- The "file opener" concept
- New config option for files: `"tiered_object=true"`, that will prevent the file's Btree from writing out blocks that step on the tiered's Btree.  Getting the config strings copied too and fro to make this work is pretty ugly, I'll probably need to add some comments or (better) find a way to clean it up.
- Btree's and Tiered dhandles are looking even more alike, there are probably some opportunities for better code sharing.

keitharnoldsmith commented on src/tiered/tiered_handle.c:
```diff
@@ -113,6 +114,15 @@ __tiered_create_local(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+    /*
+     * We should have a better way to handle the initial case of setting the object number.
+     */
+    if (((WT_BTREE *)tiered->iface.handle)->bm != NULL) {
+        WT_WITH_DHANDLE(
+          session, &tiered->iface, ret = __wt_btree_switch_object(session, tiered->current_id, 0));
+        WT_ERR(ret);
```
Instead of reaching into the btree to figure out if the block manager has been created, could we let `__wt_btree_switch_object()` handle this?  

Alternately we could add a flag on `WT_TIERED` to indicate that we are initializing. Or perhaps the inverse, a flag that says our BTREE is setup.

ddanderson replied:
I rewrote as you suggested.

keitharnoldsmith reviewed (approved):
LGTM.

One comment about possibly simplifying the initial case in `__tiered_create_local()`. 

There are a few things that feel a bit awkward, but they're already tagged and ticketed for future work.  So LGTM!

sueloverso commented on src/btree/bt_handle.c:
```diff
@@ -401,6 +398,12 @@ __btree_conf(WT_SESSION_IMPL *session, WT_CKPT *ckpt)
     else
         F_SET(btree, WT_BTREE_NO_LOGGING);
 
+    WT_RET(__wt_config_gets(session, cfg, "tiered_object", &cval));
```
Further down is the `readonly` configuration that should be removed/modified/whatever once you decide which of these we need.

sueloverso commented on src/include/block.h:
```diff
@@ -315,6 +318,19 @@ struct __wt_block_desc {
  */
 #define WT_BLOCK_DESC_SIZE 16
 
+#define WT_TIERED_TOP_OBJECT_ID 0xFFFFFFFFFFFFFFFFULL
```
I find `TOP` a confusing term. Maybe `WT_TIERED_OBJECT_ID_MAX`? Or can this just be defined as `UINT64_MAX` - does it really need its own define or could we just use one of the existing ones?

ddanderson replied:
I named this `WT_TIERED_CURRENT_ID` and added a comment, it's a way in the opener API to reference the current id.

sueloverso commented on src/tiered/tiered_opener.c:
```diff
@@ -0,0 +1,99 @@
+            WT_RET_MSG(session, EINVAL, "expected a 'file:' URI");
+    } else {
+        WT_ERR(
+          __wt_tiered_name(session, &tiered->iface, object_id, WT_TIERED_NAME_OBJECT, &object_uri));
```
The `wt_tiered_name` function could take a (new) `WT_TIERED_NAME_ONLY` flag perhaps (or `WT_TIERED_NAME_NO_PREFIX`). Callers in `conn_tiered.c:wt_tiered_do_flush` are also using it in this manner to the the URI and then stripping off the prefix. That could be another ticket, just pointing it out.

ddanderson replied:
I created WT-7635 to track this.

sueloverso reviewed (approved):
LGTM - I have a few comments.

sueloverso commented:
I took a look and the failing task is `tiered04`. That should be fixed before merge. If it was random known problem it would be okay. But `tiered04` could be real.

ddanderson commented on dist/api_data.py:
```diff
@@ -272,6 +272,10 @@ def __ge__(self, other):
+    Config('tiered_object', 'false', r'''
+        this file is a tiered object. When opened on its own, it is marked as
+        readonly and may be restricted in other ways''',
+        type='boolean', undoc=True),
```
Yes - the extra functionality currently is no checkpoint.  It's also likely that we shouldn't allow opening of this object on its own even as readonly (after all, it may have references to other objects, and the Block manager won't know what to do without viewpoint offered when opening "tiered:").  I'll remove readonly.

ddanderson replied:
As it turns out, we'll need readonly independently, as `"object:"` uris "inherit" from `"file:"` and objects need readonly.  So I'm going to leave this as is.

sueloverso replied:
Thanks for researching it.

ddanderson commented:
tiered04 doesn't fail for me, but it seems to be failing consistently with the PR builds.  Perhaps it's a difference in diagnostic vs no-diagnostic builds

ddanderson commented:
@sueloverso and @keitharnoldsmith, please take another look.  I believe I've addressed the PR comments.  The errors in tiered04 would appear to be caused by two threads both getting the dhandle exclusively.  I didn't change that, but now the worker thread that does the flush of the file retries on getting the dhandle.

sueloverso commented on src/conn/conn_tiered.c:
```diff
@@ -134,9 +134,9 @@ __tier_flush_meta(
     char *newconfig, *obj_value;
     const char *cfg[3] = {NULL, NULL, NULL};
-    bool tracking;
+    bool tracking, release;
```
Minor: alphabetize

sueloverso commented on src/conn/conn_tiered.c:
```diff
@@ -134,9 +134,9 @@ __tier_flush_meta(
+    bool tracking, release;
 
-    tracking = false;
+    tracking = release = false;
```
Alphabetize on this line too

sueloverso commented on src/conn/conn_tiered.c:
```diff
@@ -180,6 +182,7 @@ __wt_tier_do_flush(
     WT_DECL_RET;
     WT_FILE_SYSTEM *bucket_fs;
     WT_STORAGE_SOURCE *storage_source;
+    uint32_t millis, tries;
```
Minor: WT uses `msecs` or `msec` generally. And `retry`  or `retries` instead of tries.

ddanderson replied:
Fixed.

sueloverso commented on src/conn/conn_tiered.c:
```diff
@@ -194,8 +197,21 @@ __wt_tier_do_flush(
+            break;
+        __wt_sleep(0, millis * WT_THOUSAND);
+        millis *= 2;
+        /* TODO: tiered: statistic for retries */
```
I suggest adding this statistic. It is two additional lines, the definition in the `stat_data.py` file and one increment here. That doesn't seem worth a new ticket.

ddanderson replied:
Done.

sueloverso commented on src/conn/conn_tiered.c:
```diff
@@ -194,8 +197,21 @@ __wt_tier_do_flush(
+     * held by the thread that queues the flush tier work item. As a result, the handle may be busy,
+     * so retry as needed, up to a few seconds.
+     */
+    for (tries = 1, millis = 100; tries <= 6; tries++) {
```
There is a `WT_RETRY_MAX` define used in a few places that is 10. I suggest using that.

keitharnoldsmith replied:
A potential problem with letting the number of (re)tries go to 10 is that the last sleep would be 100 seconds, which seems rather long for this type of race.  Obviously you could start with a shorter sleep, or keep the current retry count.  Either way it seems like guesswork until we start using the system for real and see how often this race happens and how long it lasts.

sueloverso replied:
I would probably adjust the multiplier first. Try the 10 times at a fixed, short interval.

ddanderson replied:
Since msecs is doubling each time, I'm a little wary of bumping to 10, that makes the last wait 50 seconds.  I changed the code to exit the loop when the msecs > a few seconds, matching the comment.

sueloverso reviewed (approved):
I have a couple small comments about the new EBUSY code. But it's good to go after that.

keitharnoldsmith commented:
LGTM

ddanderson commented on src/btree/bt_handle.c:
```diff
@@ -111,14 +112,10 @@ __wt_btree_open(WT_SESSION_IMPL *session, const char *op_cfg[])
-    filename = dhandle->name;
-    if (!WT_PREFIX_SKIP(filename, "file:"))
-        WT_ERR_MSG(session, EINVAL, "expected a 'file:' URI");
+    WT_ERR(__wt_tiered_opener(session, dhandle, &opener, &filename));
```
Done.

ddanderson commented on src/block/block_tiered.c:
```diff
@@ -9,77 +9,98 @@
-    ++block->objectid;
-    WT_ERR(__wt_buf_fmt(session, tmp, "%s.%08" PRIu32, block->name, block->objectid));
-    filename = tmp->data;
+    if (object_id * sizeof(WT_FILE_HANDLE *) < block->ofh_alloc &&
```
Done.
